### ISSUE TITLE
Supported child collection formats when passed an array value to a non-array property

### ISSUE THREAD

**Issue by Nysosis** [JasperFx/marten#1105]
Marten Version: 3.0.1
Framework: DotNetCore 2.1
Postgres Version: PostgreSQL 9.6.1, compiled by Visual C++ build 1800, 64-bit

There seems to be a discrepancy on the way that collections are being serialised into the JSON document.

Given this class:
```
	public class MartenData {
		public int Id { get; set; }

		public string[] Array { get; set; }
		public IEnumerable<string> Enumerable { get; set; }
		public IEnumerable<string> IListAsArray { get; set; }
		public IEnumerable<string> IListAsList { get; set; }
		public List<string> List { get; set; }
	}
```

and this instance:
```
		public async Task Create() {
			this._Session.Store(new MartenData {
				Id = 1,
				Array = new[] { "a", "b" },
				Enumerable = new[] { "a", "b" },
				IListAsArray = new[] { "a", "b" },
				IListAsList = new List<string> { "a", "b" },
				List = new List<string> { "a", "b" }
			});

			await this._Session.SaveChangesAsync();
		}
```

The value saved into the database is:
```
{
    "IListAsList": [
        "a",
        "b"
    ],
    "List": [
        "a",
        "b"
    ],
    "IListAsArray": {
        "$type": "System.String[], System.Private.CoreLib",
        "$values": [
            "a",
            "b"
        ]
    },
    "Array": [
        "a",
        "b"
    ],
    "Enumerable": {
        "$type": "System.String[], System.Private.CoreLib",
        "$values": [
            "a",
            "b"
        ]
    },
    "Id": 1
}
```

Note on `IListAsArray`, and `Enumerable` have extended data about the type, this is resulting on `where` statements on these child properties to not work.

**Comment by oskardudycz** [comment]
You can try to setup TypeNameHandling settings of Json serializer to None https://www.newtonsoft.com/json/help/html/SerializeTypeNameHandling.htm 

I have

**Comment by oskardudycz** [comment]
@Nysosis I added tests for your case https://github.com/JasperFx/marten/pull/1107/files

Could you compare that they reflect your query and your json serializer configuration?

**Comment by oskardudycz** [comment]
@Nysosis Ok, my tests were false positive. Now they're failing. I'll look on that.

**Comment by oskardudycz** [comment]
@Nysosis ok, updated tests, now they don't have false positive results. As a solution for now, you can setup JsonSerializer options to use `TypeNameHandling.None`. See [sample in my tests](https://github.com/JasperFx/marten/pull/1107/files#diff-69c458cf4968cb9204c17b7628660b15R105)

@jeremydmiller do you remember why we don't have it setup by default? I see that there is logic of using "clean" serializer for queries, but it doesn't work in this case. Is it a bug?

**Comment by Nysosis** [comment]
Thanks @oskardudycz that looks to cover all the bases.
I'm happy to use `TypeNameHandling` for now. Be interesting to see if there's a reason it can't be the default behaviour though.

**Comment by oskardudycz** [comment]
@Nysosis just be carefull, `TypeNameHandling.None` will not work if you're using interfaces instead of the exact types (eg. instead type `User` you have property with interface `IUser`)

**Comment by oskardudycz** [comment]
@mysticmind I'll address that in my PR.

**Comment by mysticmind** [comment]
In retrospect, it would be best practice not to use `IEnumerable` for storage of data since the serializer will need to add additional metadata to infer proper type to serialize or deserialize an object.

We will need to add some documentation to suggest users not to use `IEnumerable` properties for Marten document storage. 

@oskardudycz Please let me know your thoughts, we can plan to close this issue.

**Comment by oskardudycz** [comment]
Using interfaces, so `IEnumerable`, `IReadonlyCollection`, `IList`, `ICollection` is quite common practice. Imho our mechanism of using two types of selector (raw and typed) is also not ideal. I'll try too get back to my open PR and check what are the options.
For sure we need to update documentation, to clarify this consfusion, as it's quite common question. I'll do that in my PR.

**Comment by mysticmind** [comment]
This issue is also related to handling storage of properties as array using the custom JsonConverter. Solution proposed in #1192 will solve this issue as well.

**Comment by jeremydmiller** [comment]
Closed by #1107
